dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageChecker.java

```diff
@@ -182,6 +187,90 @@ private static void printStaticLinkageErrors(
     return jarPaths.build();
   }
 
+  /**
+   * Finds linkage errors in the input classpath and generates a static linkage check report.
+   */
+  StaticLinkageCheckReport findLinkageErrors() {
+    ImmutableList<Path> jarFilePaths = classDumper.getInputClasspath();
+
+    ImmutableMap.Builder<Path, SymbolTable> jarToSymbols = ImmutableMap.builder();
+    for (Path jarPath : jarFilePaths) {
+      jarToSymbols.put(jarPath, scanExternalSymbolTable(jarPath));
+    }
+
+    // Validate linkage error of each reference
+    ImmutableList.Builder<JarLinkageReport> jarLinkageReports = ImmutableList.builder();
+    for (Map.Entry<Path, SymbolTable> entry : jarToSymbols.build().entrySet()) {
+      jarLinkageReports.add(findInvalidReferences(entry.getKey(), entry.getValue()));
+    }
+
+    if (reportOnlyReachable) {
+      // TODO: Optionally, report errors only reachable from entry point classes
+    }
+
+    return StaticLinkageCheckReport.create(jarLinkageReports.build());
+  }
+
+  private JarLinkageReport findInvalidReferences(Path jarPath, SymbolTable jarSymbolTable) {
+    JarLinkageReport.Builder reportBuilder = JarLinkageReport.builder().setJarPath(jarPath);
+
+    // TODO(suztomo): implement validation for field, method and class references in the table
+    return reportBuilder.build();
+  }
+
+  /**
+   * Scans class files in the jar file and creates {@link SymbolTable} populated with class names
+   * defined within the jar and symbolic references to other classes outside the jar.
+   *
+   * @param jarFilePath absolute path to a jar file.
+   * @return Table containing symbol references and classes defined in the jar file.
+   */
+  static SymbolTable scanExternalSymbolTable(Path jarFilePath) {
+    SymbolTable.Builder symbolTableBuilder = SymbolTable.builder();
+    ImmutableSet.Builder<String> symbolTableClassNameBuilder = ImmutableSet.builder();
+    ImmutableSet.Builder<ClassSymbolReference> classSymbolReferences = ImmutableSet.builder();
+    ImmutableSet.Builder<MethodSymbolReference> methodSymbolReferences = ImmutableSet.builder();
+    ImmutableSet.Builder<FieldSymbolReference> fieldSymbolReferences = ImmutableSet.builder();
+
+    String pathToJar = jarFilePath.toString();
+    SyntheticRepository repository = SyntheticRepository.getInstance(new ClassPath(pathToJar));
+
+    try {
+      for (JavaClass javaClass : ClassDumper.topLevelJavaClassesInJar(jarFilePath, repository)) {
+        SymbolTable partialSymbolTable = ClassDumper.scanClassSymbolTable(javaClass);
+
+        classSymbolReferences.addAll(partialSymbolTable.getClassReferences());
```

Comment:
If it makes sense, you can add a concrete method to `SymbolTable.Builder` that takes a `SymbolTable` and merges it, like `Builder merge(SymbolTable other)`. You can have any concrete methods you want on an AutoValue builder (or the class itself).